### PR TITLE
RichHandler implementation 2: fancy tracebacks

### PR DESCRIPTION
**Pull request by SajidAlamQB**
Signed-off-by: SajidAlamQB <90610031+SajidAlamQB@users.noreply.github.com>

## Description

This is the second part to the RichHandler implementation. In this PR we will look to adopt the [rich exception handler](https://www.willmcgugan.com/blog/tech/post/better-python-tracebacks-with-rich/). 

### rich_tracebacks vs rich.traceback.install
The difference between setting [rich_tracebacks: True](https://rich.readthedocs.io/en/stable/logging.html#handle-exceptions) on our rich logging handler vs. [rich.traceback.install](https://rich.readthedocs.io/en/stable/traceback.html#traceback-handler) is the former doesn't effect every traceback. During testing I had to retrieve the rich logger via `log = logging.getLogger("rich")` and raise excpetions through `log.excpetion("blah")` to get a rich traceback. With [rich.traceback.install] every traceback was already turned into a rich output.

From the rich backend, both methods make use of the [Traceback class](https://rich.readthedocs.io/en/stable/_modules/rich/traceback.html#Traceback), which helps render the tracebacks. However, [rich.traceback.install] [install](https://rich.readthedocs.io/en/stable/_modules/rich/traceback.html#install) method does some more lower-level work to replace all tracebacks.

We may want to go with the second option for the conversion of all Tracebacks and not just exceptions. Would love to get your thoughts on this.

## Development notes

After some discussion with @AntonyMilneQB we decided it would be best to go with the second option and have rich.traceback.install be the way rich traceback is implemented.

To disable rich traceback in a project's `settings.py` you can `import sys` and add the following line: `sys.excepthook = sys.__excepthook__`. This will revert the exception handler.

Rich `install` allows us to suppress certain external tracebacks, with this in mind we will suppress the `click` library. We won't suppress `kedro` as this would make debugging too difficult.

Rich traceback on windows e2e tests fails as output can't be decoded by utf-8 encoding. To resolve this we need to set `PYTHONIOENCODING=utf-8` this environment variable.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro/pull/1603"></a>

### CODE REVIEW

**Comment by AntonyMilneQB** [comment]
At a glance this looks great! A few more things I think we should do here:
* Add a comment to the code explaining why we suppress `click` but not `kedro` and what suppressing `shutil.which("kedro")` does
* Do a little screenshot or two in this PR just to show what it looks like
* Check what it does in ipython and jupyter
* Add to release notes

**Comment by SajidAlamQB** [comment]
Rich Traceback:

Tracebacks

**Comment by SajidAlamQB** [comment]
Kedro rich on Windows CLI:

**Review by MerelTheisenQB — APPROVED** [review]
These changes make sense and the output looks great! 👍 I just had one small clarification question, but not blocking.

File: kedro/config/default_logger.py [hunk 1]
```diff
@@ -4,12 +4,20 @@
...
 CURRENT_DIR = os.path.dirname(__file__)
 
 with open(os.path.join(CURRENT_DIR, "logging.yml"), encoding="utf-8") as conf_file:
     LOGGING_CONFIG = yaml.safe_load(conf_file.read())
     logging.config.dictConfig(LOGGING_CONFIG)
     logging.captureWarnings(True)
+
+# We suppress click here to hide tracebacks related to it conversely,
+# kedro is not suppressed to show its tracebacks for easier debugging.
+# shutil is used to get the kedro executable path to hide the top level traceback.
```
[the remark this answers is not in the document]
Comment (SajidAlamQB) [reply kedro/config/default_logger.py:22]: The top-level refers to this: 
top level traceback

When we use `shutil.which('kedro')` we can suppress that top header which isn't useful for debugging:
suppressed top level

This is just to make it cleaner to read the traceback.

[the remark this answers is not in the document]
Comment (MerelTheisenQB) [reply kedro/config/default_logger.py:22]: Thanks for clarifying! That makes total sense 👍

**Review by AntonyMilneQB — APPROVED** [review]
This looks great! Especially nice to get that `shutil.which` thing figure out 👏 

I've had a good play around and all seems to work well. The tracebacks are a little too long at the moment for my personal taste, but that doesn't really matter. I wonder whether we should consider not doing `show_locals` or suppressing kedro in the future (but this would potentially make it harder for us to debug stuff). This is just my individual taste though, so I'm very happy to release like this and let's see what users make of the current options and whether there's any feedback on it!

FYI I just added a line to the PyCharm docs.

**Pull request merged**